### PR TITLE
WIP - Feature threejs r103

### PR DESCRIPTION
**Pull request by kalwalt**
**What kind of change does this PR introduce?**
This PR upgrade the Three.js version (r103) in the examples and in other places. it upgrade some deprecated method ( note this is not directly related to Three.js):

```diff
- onResize()
+ onResizeElement()
- copySizeTo 
+ copyElementSizeTo()

```
and in some examples and in threex-armarkerhelper.js:

```diff
+ THREE.AxesHelper()
- THREE.AxisHelper()
```

**Can it be referenced to an Issue? If so what is the issue#  ?**
see issue https://github.com/jeromeetienne/AR.js/issues/513 and https://github.com/jeromeetienne/AR.js/issues/511

**How can we test it?**

Tested in Ubuntu 18.04 and Android. It willl be nice if other developers can test it in other OS.
**Summary**

The Three.js version included is very old, this issue solve some issue related to Three.js version and make possible to benefit of the new features and bug fixes of the lib. 
**Does this PR introduce a breaking change?**
I hope no.
**Other information**
Maybe i will add a new example with some new Three.js feature.

### CODE REVIEW

**Comment by kalwalt** [comment]
The above commits are not connected to the Three.js library but I thought it useful to upgrade, Note that the `dev` example already uses these methods:
https://github.com/jeromeetienne/AR.js/blob/4f25721bd3df595a952e030e7afcbd7e837c58d9/three.js/examples/dev.html#L88

**Comment by kalwalt** [comment]
In the console i receive also this warning:

`THREE.AxisHelper has been renamed to THREE.AxesHelper.`

I think i colud make also these changes, renaming:

```diff
- THREE.AxisHelper 
+ THREE.AxesHelper.
```

it is located 29 times in 27 files, but it worth to change them.

**Comment by kalwalt** [comment]
i have this issue: i builded the libs with the makefile commmand  when i run the basic aframe example
(Ar.js/aframe/examples/basic.html) i get this message:

```bash
A-Frame Version: 0.6.1 (Date 29-07-2017, Commit #1bfc562) aframe.min.js:41:18046
three Version: ^0.86.0 aframe.min.js:41:18119
WebVR Polyfill Version: ^0.9.36 aframe.min.js:41:18175
extras:primitives:warn Mapping keys should be specified in lower case. The mapping key minConfidence may not be recognized aframe.min.js:2:19179
extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothCount may not be recognized aframe.min.js:2:19179
extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothTolerance may not be recognized aframe.min.js:2:19179
extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothThreshold may not be recognized aframe.min.js:2:19179
extras:primitives:warn Mapping keys should be specified in lower case. The mapping key hit-testing-renderDebug may not be recognized aframe.min.js:2:19179
Live reload enabled. basic.html:60:4
Successfully compiled asm.js code (total compilation time 207ms) aframe-ar.js
THREE.WebGLRenderer 86 aframe.min.js:15:19494
AR.js 1.6.2 - trackingBackend: artoolkit aframe-ar.js:7139:2
Allocated videoFrameSize 1228800 aframe-ar.js:2:23001
Pattern detection mode set to 1. aframe-ar.js:2:23001
Pattern ratio size set to 0.500000. aframe-ar.js:2:23001
L’utilizzo di mozImageSmoothingEnabled è deprecato. Al suo posto utilizzare la proprietà senza prefisso imageSmoothingEnabled. aframe-ar.js:5435:2
ARjs.Anchor - changeMatrixMode: modelViewMatrix / markersAreaEnabled: true aframe-ar.js:6599:2
```

look the Three.js version printed in the console is 86 !  It should be 103...

Instead the Three.js example (Ar.js/three.js/examples/basic.html):

```bash
THREE.WebGLRenderer 103 three.min.js:184:454
Live reload enabled. basic.html:191:4
Successfully compiled asm.js code (total compilation time 242ms) ar.js
Allocated videoFrameSize 1228800 ar.js:2:23001
Pattern detection mode set to 1. ar.js:2:23001
Pattern ratio size set to 0.500000. ar.js:2:23001
L’utilizzo di mozImageSmoothingEnabled è deprecato. Al suo posto utilizzare la proprietà senza prefisso imageSmoothingEnabled. ar.js:5435:2
```

i get the actual version 103.
I have the suspect that the aframe in use needs r86? or there are other changes required in the code ?
@nicolocarpignoli @janpio what do you think?

**Comment by kalwalt** [comment]
using the aframe.js i arrived to this conclusion:
it point to this line:

```javascript
// at line 26036
function WebGLRenderer( parameters ) {

		console.log( 'THREE.WebGLRenderer', REVISION 
);
```
and the constant is set as:

```javascript
// at linline 5919
var REVISION = '86';
```
so aframe  require at least revision 86? and maybe i should update aframe to a newer version of it.

**Comment by kalwalt** [comment]
My initial thought was that Ar.js embedd his own Three.js version but it is  not right, aframe has already embedded his own Three.js version. So if i  updated the Three.js version and the ThreeX source (see also the Three.axesHelper error in a previuos comment), i should use the latest version of aframe that support the r103. I will try this way.

**Comment by kalwalt** [comment]
with the latest from aframe:

```bash
[Deprecation] document.registerElement is deprecated and will be removed in M73, around March 2019. Please use window.customElements.define instead. See https://www.chromestatus.com/features/4642138092470272 for more details.
module.exports.registerElement @ aframe-v0.9.2.min.js:259
aframe-v0.9.2.min.js:353 A-Frame Version: 0.9.2 (Date 2019-05-06, Commit #f57a1fa)
aframe-v0.9.2.min.js:353 three Version (https://github.com/supermedium/three.js): ^0.102.2
aframe-v0.9.2.min.js:353 WebVR Polyfill Version: ^0.10.10
aframe-v0.9.2.min.js:95 THREE.WebGLRenderer 103dev
aframe-v0.9.2.min.js:23 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key minConfidence may not be recognized 
aframe-v0.9.2.min.js:23 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothCount may not be recognized 
aframe-v0.9.2.min.js:23 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothTolerance may not be recognized 
aframe-v0.9.2.min.js:23 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothThreshold may not be recognized 
aframe-v0.9.2.min.js:23 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key hit-testing-renderDebug may not be recognized 
basic.html:60 Live reload enabled.
aframe-ar.js:7139 AR.js 1.6.2 - trackingBackend: artoolkit
aframe-ar.js:6599 ARjs.Anchor - changeMatrixMode: modelViewMatrix / markersAreaEnabled: true
aframe-ar.js:2 Allocated videoFrameSize 1228800
aframe-ar.js:2 Pattern detection mode set to 1.
aframe-ar.js:2 Pattern ratio size set to 0.500000.
```
Three.axesHelper error of course disappear see the Three.js version 0.102.2
i could downgrade the Three.js version to match them, I will think about it...

**Comment by kalwalt** [comment]
I notice more warnings in the console, this is what i get from master:

```javascript
extras:primitives:warn Mapping keys should be specified in lower case. The mapping key minConfidence may not be recognized 
aframe.js:3217 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key hit-testing-renderDebug may not be recognized 

```
and this is from the branch i'm working (synced with dev):

```javascript
extras:primitives:warn Mapping keys should be specified in lower case. The mapping key minConfidence may not be recognized 
aframe-v0.9.2.js:2396 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothCount may not be recognized 
aframe-v0.9.2.js:2396 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothTolerance may not be recognized 
aframe-v0.9.2.js:2396 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key smoothThreshold may not be recognized 
aframe-v0.9.2.js:2396 extras:primitives:warn Mapping keys should be specified in lower case. The mapping key hit-testing-renderDebug may not be recognized
```
Have you an idea where they come from?

**Comment by kalwalt** [comment]
And also these Violation messages:


```javascript
aframe-v0.9.2.js:75381 [Violation] Added non-passive event listener to a scroll-blocking 'touchmove' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
aframe-v0.9.2.js:66763 [Violation] Added non-passive event listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
addEventListeners @ aframe-v0.9.2.js:66763
update @ aframe-v0.9.2.js:66695
initComponent @ aframe-v0.9.2.js:73768
updateProperties @ aframe-v0.9.2.js:73740
updateComponent @ aframe-v0.9.2.js:72422
updateComponents @ aframe-v0.9.2.js:72395
entityLoadCallback @ aframe-v0.9.2.js:72181
emitLoaded @ aframe-v0.9.2.js:73109
aframe-v0.9.2.js:66763 [Violation] Added non-passive event listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
addEventListeners @ aframe-v0.9.2.js:66763
play @ aframe-v0.9.2.js:66707
play @ aframe-v0.9.2.js:74225
initComponent @ aframe-v0.9.2.js:73772
updateProperties @ aframe-v0.9.2.js:73740
updateComponent @ aframe-v0.9.2.js:72422
updateComponents @ aframe-v0.9.2.js:72395
entityLoadCallback @ aframe-v0.9.2.js:72181
emitLoaded @ aframe-v0.9.2.js:73109
dev.html:93 Live reload enabled.
aframe-v0.9.2.js:66763 [Violation] Added non-passive event listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
```

complete [gist](https://gist.github.com/kalwalt/592b077d0cc1d16bf28aee6f4557d84f)

**Comment by nicolocarpignoli** [comment]
the 'lower-case' warning may be related to  https://github.com/jeromeetienne/AR.js/pull/528?

**Comment by kalwalt** [comment]
> the 'lower-case' warning may be related to #528?

yes, @nicolocarpignoli  it is related! it seems that is completely solved, but I did only a quick test, will test further today.

**Comment by kalwalt** [comment]
@nicolocarpignoli one note: in the aframe folder https://github.com/jeromeetienne/AR.js/tree/master/aframe/examples/vendor/aframe/build there are different files (old) we should keep them?  for example https://github.com/jeromeetienne/AR.js/blob/master/aframe/examples/vendor/aframe/build/aframe-v0.6.1-three-r86.js 
and https://github.com/jeromeetienne/AR.js/blob/master/aframe/examples/vendor/aframe/build/aframe.js
seems the same, they output same version and commit: 

```js
A-Frame Version: 0.6.1 (Date 29-07-2017, Commit #1bfc562)
aframe.js:77533 three Version: ^0.86.0

```
In my opinion if there are no reasonable questions we could remove those old files.
I am also testing with Android in localhost, i willl report my result at the end of the day or at least tomorrow.

**Comment by nicolocarpignoli** [comment]
@kalwalt any news?

**Comment by kalwalt** [comment]
> @kalwalt any news?

@nicolocarpignoli i will do a quicly re-check this afternoon.

File: three.js/examples/measure-it.html [hunk 1]
```diff
@@ -189,7 +189,7 @@
 		geometry.verticesNeedUpdate = true
 
 		geometry.computeBoundingSphere();
-		geometry.computeLineDistances();
```
Comment (kalwalt) [inline three.js/examples/measure-it.html]: This is deprecated in Three.js r103, instead we should use THREE.Line.computeLineDistances()

**Comment by kalwalt** [comment]
An issue also with trackingbackend-switch.html example:
```js
threex-artoolkitsource.js:315 Assertion failed: console.assert
ARjs.Source.onResizeElement @ threex-artoolkitsource.js:315
onReady @ trackingbackend-switch.html:118
onSourceReady @ threex-artoolkitsource.js:83
(anonymous) @ threex-artoolkitsource.js:223
setInterval (async)
success @ threex-artoolkitsource.js:221
Promise.then (async)
(anonymous) @ threex-artoolkitsource.js:210
Promise.then (async)
ARjs.Source._initSourceWebcam @ threex-artoolkitsource.js:185
ARjs.Source.init @ threex-artoolkitsource.js:65
(anonymous) @ trackingbackend-switch.html:117
```

**Comment by kalwalt** [comment]
That message is normal, the torus knot iwill disappear when  a marker will be detected ( Hiro or Aruco or a Tango device) , I also updated another deprecated function https://github.com/jeromeetienne/AR.js/pull/515/commits/7a7da756ce2c74aecd5c461822336bc6443643d2

**Comment by kalwalt** [comment]
`trackingbackend-switch.html` does not works as in the master branch, i don't know why...
With the hiro marker the mesh ( the cube and the torus knot) should be attached to the marker position but instead it simply disappear. Investigating...

**Comment by kalwalt** [comment]
To complete the infos: The `console.assert` does not raise in the master branch, and in both i receive also:
```js
threex-artoolkitprofile.js:145 stop profile.trackingBackend() obsolete function. use .trackingMethod instead
ARjs.Profile.trackingBackend @ threex-artoolkitprofile.js:145
(anonymous) @ trackingbackend-switch.html:108
trackingbackend-switch.html:320 
```
I will try to updgrade that deprecated method

**Comment by kalwalt** [comment]
I arrvived to this conclusion, this is caused by some latest in this PR,:sob:  i tested also the dev branch and it works as it should.

**Comment by kalwalt** [comment]
Sorry the trackingBackend needs the onResize() method https://github.com/jeromeetienne/AR.js/blob/ee1129d26f9ffe4cee747e01fe57630730d48775/three.js/src/threex/threex-artoolkitsource.js#L396-L443, so nothing to update here!

**Comment by kalwalt** [comment]
i tested most of the examples, at least those i can,  but i this issue with minecraft demo:

```js
aframe-ar.js:5047 Uncaught TypeError: THREE.AxesHelper is not a constructor
    at new THREEx.ArMarkerHelper (aframe-ar.js:5047)
    at aframe-ar.js:6657
    at Array.forEach (<anonymous>)
    at new ARjs.Anchor (aframe-ar.js:6655)
    at aframe-ar.js:8391
THREEx.ArMarkerHelper @ aframe-ar.js:5047
(anonymous) @ aframe-ar.js:6657
ARjs.Anchor @ aframe-ar.js:6655
(anonymous) @ aframe-ar.js:8391
setInterval (async)
init @ aframe-ar.js:8337
updateProperties @ component.js:263
module.exports.Component @ component.js:50
i @ component.js:450
value @ a-entity.js:377
value @ a-entity.js:544
e @ a-entity.js:513
value @ a-entity.js:495
(anonymous) @ a-entity.js:306
(anonymous) @ a-node.js:117
Promise.then (async)
value @ a-node.js:115
value @ a-entity.js:302
value @ a-entity.js:94
value @ a-register-element.js:150
value @ a-register-element.js:150
```
This is related to multimarker, I will test also this but i have to print, i will do tomorrow, if i have time. Maybe THREE.Axes is very different from THREE.Axis ?

it fails in specific at Array.forEach:

https://github.com/jeromeetienne/AR.js/blob/ee1129d26f9ffe4cee747e01fe57630730d48775/three.js/src/new-api/arjs-anchor.js#L83

**Comment by kalwalt** [comment]
@nicolocarpignoli My final thought: I can still look at this last problem (very specific), but in my opinion the code should works 98%; on the other side many example files should probably be reviewed, some are very old, they don't work properly even in the master branch. I don't think the updated versions of three.js and aframe.js cause such radical code breaks. All this other work, however, should be done in another PR and if problems arise, fix them with specific PR.

**Comment by kalwalt** [comment]
reading this [commit](https://github.com/mrdoob/three.js/commit/845999e06fb8c64c30d413bf668aa9da68357e01) THREE.Axes has no changes in the code, the only relevant change is in the name.
I have to re-check the AR.js code, maybe i miss something out.

**Comment by kalwalt** [comment]
BIngo! :rofl: :+1:  The minecraft example use a very old aframe version `0.6.0` before the THREE.Axis -> THREE.Axes change:
https://github.com/jeromeetienne/AR.js/blob/ee1129d26f9ffe4cee747e01fe57630730d48775/aframe/demos/demo-minecraft-launch-demo/minecraft.html#L3
using aframe 0.9.0, that one i included solved the isssue. I will make the changes in the code where i found old version of aframe.

**Comment by kalwalt** [comment]
i left the https import for the minecraft examples but if you want i can change using the embedded aframe version, i mean that one in this [folder](https://github.com/jeromeetienne/AR.js/tree/dev/aframe/examples/vendor/aframe/build) 
Another issue arise while upgrading aframe with the demo-hatsune-miku, probably a bit outdated example, infact aframe reccomend the [gltf](https://aframe.io/docs/0.9.0/introduction/models.html) approach.
I will look at this tomorrow if i can. :smile:

**Comment by kalwalt** [comment]
@nicolocarpignoli The `demo-hatsune-miku` is based on the takhirox [a-mmd](https://github.com/takahirox/a-mmd) loader. It does not receive any update by 2 years and the most recent version of it, does not support the newest aframe version, so why we should keep those demos alive? i propose to delete all the `demo-hatsune-miku` and in a near future make another alternative example, maybe with  the Three.js MMDLoader or another with GLTF standard (reccomended by aframe).

**Comment by kalwalt** [comment]
i suggest an example like this one: https://github.com/kalwalt/kalwalt-interactivity-AR/blob/master/aframe-ar-gltf.html it is based on the [CesiumMan  gltf example from Three.js](https://threejs.org/examples/#webgl_loader_gltf_extensions), It use [aframe-extras](https://github.com/donmccurdy/aframe-extras) to load the animations. But we could add even a simpler example without the need of aframe-extras.

**Comment by kalwalt** [comment]
with the Duck model: https://github.com/kalwalt/kalwalt-interactivity-AR/blob/master/aframe-ar-simple-gltf.html

**Comment by nicolocarpignoli** [comment]
Hi @kalwalt, I see no big problems with this PR, only thing left to solve are conflicts. Can you do that so we can think then of merging this one

**Comment by nicolocarpignoli** [comment]
also, I see file modified under /build folder. I guess, you build the lib only for test purposes right? 

So I will not review them. 
They will eventually change again when we will release next ar.js version

**Comment by kalwalt** [comment]
> Hi @kalwalt, I see no big problems with this PR, only thing left to solve are conflicts. Can you do that so we can think then of merging this one

yes, i will solve them. No problem.

I stopped a bit with this PR because i was waiting you response on my previuos comment/question about a-mmd loader and examples that will not work with the new version; see the comment in the quote above: 

> @nicolocarpignoli The `demo-hatsune-miku` is based on the takhirox [a-mmd](https://github.com/takahirox/a-mmd) loader. It does not receive any update by 2 years and the most recent version of it, does not support the newest aframe version, so why we should keep those demos alive? i propose to delete all the `demo-hatsune-miku` and in a near future make another alternative example, maybe with the Three.js MMDLoader or another with GLTF standard (reccomended by aframe).

and regarding your last comment:

> also, I see file modified under /build folder. I guess, you build the lib only for test purposes right?
> 
> So I will not review them.
> They will eventually change again when we will release next ar.js version

They are modified because i did some changes in [threex-armarkerhelper.js](https://github.com/jeromeetienne/AR.js/pull/515/files#diff-ae22f8ada86b8504cb61d1a420fd1371) regarding the THREE.AxisHelper() issue. i need to rebuild the libs otherwise the libs will be not updated and aligned to that changes.

**Comment by nicolocarpignoli** [comment]
thank you a lot!

**Comment by nicolocarpignoli** [comment]
@kalwalt can you point me the obsolete examples? I will remove them. They actually are more dangerous than helpful.

**Comment by kalwalt** [comment]
Wow my First PR here! 
@nicolacarpignoli. I will make a list. ☺️

**Comment by nicolocarpignoli** [comment]
Ok thx! BTW, it's nicolocarpignoli 😊

**Comment by kalwalt** [comment]
Sorry It was a typo!

**Comment by kalwalt** [comment]
@nicolocarpignoli the examples are in https://github.com/jeromeetienne/AR.js/tree/master/aframe/demos/demo-hatsune-miku

**Pull request merged**